vg/vgpdf/vgpdf.go

```diff
@@ -25,123 +29,230 @@ const DPI = 72
+	c.pdf.SetMargins(0, 0, 0)
+	c.pdf.AddPage()
+	c.Push()
+	c.Translate(vg.Point{0, h})
+	c.Scale(1, -1)
+
 	return c
 }
 
+func (c *Canvas) DPI() float64 {
+	return float64(c.dpi)
+}
+
+func (c *Canvas) cur() *context {
+	return &c.ctx[len(c.ctx)-1]
+}
+
 func (c *Canvas) Size() (w, h vg.Length) {
 	return c.w, c.h
 }
 
 func (c *Canvas) SetLineWidth(w vg.Length) {
-	c.page.SetLineWidth(unit(w))
+	c.cur().linew = w
+	lw := c.unit(w)
+	c.pdf.SetLineWidth(lw)
 	c.lineVisible = w > 0
 }
 
 func (c *Canvas) SetLineDash(dashes []vg.Length, offs vg.Length) {
-	ds := make([]pdf.Unit, len(dashes))
+	ds := make([]float64, len(dashes))
 	for i, d := range dashes {
-		ds[i] = unit(d)
+		ds[i] = c.unit(d)
 	}
-	c.page.SetLineDash(unit(offs), ds)
+	c.pdf.SetDashPattern(ds, c.unit(offs))
 }
 
 func (c *Canvas) SetColor(clr color.Color) {
-	c.page.SetStrokeColor(pdfColor(clr))
-	c.page.SetColor(pdfColor(clr))
+	if clr == nil {
+		clr = color.Black
+	}
+	c.cur().drawc = clr
+	c.cur().fillc = clr
+	c.pdf.SetFillColor(rgb(clr))
+	c.pdf.SetDrawColor(rgb(clr))
+	c.pdf.SetTextColor(rgb(clr))
 }
 
 func (c *Canvas) Rotate(r float64) {
-	c.page.Rotate(float32(r))
+	c.pdf.TransformRotate(-r*180/math.Pi, 0, 0)
 }
 
 func (c *Canvas) Translate(pt vg.Point) {
-	c.page.Translate(unit(pt.X), unit(pt.Y))
+	xp, yp := c.pdfPoint(pt)
+	c.pdf.TransformTranslate(xp, yp)
 }
 
 func (c *Canvas) Scale(x float64, y float64) {
-	c.page.Scale(float32(x), float32(y))
+	c.pdf.TransformScale(x*100, y*100, 0, 0)
 }
 
 func (c *Canvas) Push() {
-	c.page.Push()
+	c.ctx = append(c.ctx, *c.cur())
+	c.pdf.TransformBegin()
 }
 
 func (c *Canvas) Pop() {
-	c.page.Pop()
+	c.pdf.TransformEnd()
+	c.ctx = c.ctx[:len(c.ctx)-1]
 }
 
 func (c *Canvas) Stroke(p vg.Path) {
 	if c.lineVisible {
-		c.page.Stroke(pdfPath(c, p))
+		c.pdfPath(p, "D")
 	}
 }
 
 func (c *Canvas) Fill(p vg.Path) {
-	c.page.Fill(pdfPath(c, p))
+	c.pdfPath(p, "F")
 }
 
 func (c *Canvas) FillString(fnt vg.Font, pt vg.Point, str string) {
-	t := new(pdf.Text)
-	t.SetFont(fnt.Name(), unit(fnt.Size))
-	t.NextLineOffset(unit(pt.X), unit(pt.Y))
-	t.Text(str)
-	c.page.DrawText(t)
+	c.font(fnt, pt)
+	c.pdf.SetFont(fnt.Name(), "", c.unit(fnt.Size))
+
+	c.Push()
+	defer c.Pop()
+	c.Translate(pt)
+	// go-fpdf uses the top left corner as origin.
+	c.Scale(1, -1)
+	left, top, right, bottom := c.sbounds(fnt, str)
+	w := right - left
+	h := bottom - top
+	margin := c.pdf.GetCellMargin()
+
+	c.pdf.MoveTo(-left-margin, top)
+	c.pdf.CellFormat(w, h, str, "", 0, "BL", false, 0, "")
+}
+
+func (c *Canvas) sbounds(fnt vg.Font, txt string) (left, top, right, bottom float64) {
+	_, h := c.pdf.GetFontSize()
+	d := c.pdf.GetFontDesc("", "")
+	if d.Ascent == 0 {
+		// not defined (standard font?), use average of 81%
+		top = 0.81 * h
+	} else {
+		top = -float64(d.Ascent) * h / float64(d.Ascent-d.Descent)
+	}
+	return 0, top, c.pdf.GetStringWidth(txt), top + h
+}
+
+// whitespace determines if a rune is whitespace
+func whitespace(r rune) bool {
```

Comment:
These are not the only whitespace characters. Is there not a whitespace check function in "utf8"?